go-controller/pkg/network-attach-def-controller/network_attach_def_controller.go

```diff
@@ -66,496 +43,199 @@ type NetworkControllerManager interface {
-			return err
+		key, err := cache.MetaNamespaceKeyFunc(nad)
+		if err != nil {
+			klog.Errorf("%s: failed to sync %v: %v", nadController.name, nad, err)
+			continue
+		}
+		err = nadController.syncNAD(key, nad)
+		if err != nil {
+			return fmt.Errorf("%s: failed to sync %s: %v", nadController.name, key, err)
 		}
 	}
 
-	return nadController.ncm.CleanupDeletedNetworks(nadController.getAllNetworkControllers())
-}
-
-func (nadController *NetAttachDefinitionController) worker() {
-	for nadController.processNextWorkItem() {
-	}
-}
-
-func (nadController *NetAttachDefinitionController) processNextWorkItem() bool {
-	key, quit := nadController.queue.Get()
-	if quit {
-		return false
-	}
-	defer nadController.queue.Done(key)
-
-	err := nadController.sync(key.(string))
-	nadController.handleErr(err, key)
-	return true
+	return nil
 }
 
 func (nadController *NetAttachDefinitionController) sync(key string) error {
 	startTime := time.Now()
+	klog.V(5).Infof("%s: sync NAD %s", nadController.name, key)
+	defer func() {
+		klog.V(4).Infof("%s: finished syncing NAD %s, took %v", nadController.name, key, time.Since(startTime))
+	}()
+
 	namespace, name, err := cache.SplitMetaNamespaceKey(key)
 	if err != nil {
-		return err
+		klog.Errorf("%s: failed splitting key %s: %v", nadController.name, key, err)
+		return nil
 	}
-	klog.V(5).Infof("%s: Sync net-attach-def %s/%s", nadController.name, namespace, name)
-	defer func() {
-		klog.V(4).Infof("%s: Finished syncing net-attach-def %s/%s: %v", nadController.name, namespace, name, time.Since(startTime))
-	}()
 
 	nad, err := nadController.netAttachDefLister.NetworkAttachmentDefinitions(namespace).Get(name)
 	if err != nil && !apierrors.IsNotFound(err) {
 		return err
 	}
 
-	if nad == nil {
-		return nadController.DeleteNetAttachDef(key)
-	} else {
-		return nadController.AddNetAttachDef(nadController.ncm, nad, true)
-	}
+	return nadController.syncNAD(key, nad)
 }
 
-func (nadController *NetAttachDefinitionController) handleErr(err error, key interface{}) {
-	ns, name, keyErr := cache.SplitMetaNamespaceKey(key.(string))
-	if keyErr != nil {
-		klog.ErrorS(err, "Failed to split meta namespace cache key", "key", key)
-	}
-
-	if errors.Is(err, ErrNetworkControllerTopologyNotManaged) {
-		klog.V(2).Infof("%s: Dropping net-attach-def %q out of the queue. No network controller to manage it.", nadController.name, key)
-		err = nil
-	}
-
-	if err == nil {
-		nadController.queue.Forget(key)
-		return
-	}
-
-	if nadController.queue.NumRequeues(key) < maxRetries {
-		nadController.queue.AddRateLimited(key)
-		klog.V(2).InfoS("Error syncing net-attach-def, retrying", "net-attach-def", klog.KRef(ns, name), "err", err)
-		return
-	}
-
-	klog.Warningf("%s: Dropping net-attach-def %q out of the queue: %v", nadController.name, key, err)
-	nadController.queue.Forget(key)
-	utilruntime.HandleError(err)
-}
+func (nadController *NetAttachDefinitionController) syncNAD(key string, nad *nettypes.NetworkAttachmentDefinition) error {
```

Comment:
this function is a bit confusing to me, as I can't figure out which NAD changes are valid and what kind of cleanup is required in case of error.  Some comments would be useful

Reply:
What would you like to see? This is just the reconcile of the NAD controller, there is ample comment throughout the implementation already

Reply:
So lets read through those comments.

We have an `oldnetwork` if:
`// the NAD refers to a different network than before`
or
`// the NAD is the only NAD referring to an existing incompatible network`
or
`// the NAD refers to an existing incompatible network referred by other NADs` and `// the NAD refers to the same network as before but with different incompatible configuration`

We have an `ensureNetwork` if
`// the NAD refers to a new network`
or
`// the NAD refers to an existing compatible network`
or
`// the NAD is the only NAD referring to an existing incompatible network`

We have an error if
`// the NAD refers to an existing incompatible network referred by other NADs`

For an `oldNetwork` we
```
// remove the NAD reference from the old network and delete the network if
// it is no longer referenced
```

For an `ensureNetwork` we
```
// ensure the network associated with the NAD
```

Let's work on how to clarify this better.

Reply:
And `imcompatible` right now is `Equals` or the old `CompareNetInfo`

Reply:
here are my comment suggestions:
- for `NetAttachDefinitionController`
NAD contains a network definition by specifying network name and network params in its spec. NAD is namespace-scoped, while network is cluster-scoped, which means that multiple NADs can use the same network. NetAttachDefinitionController handles NAD events and manages networks defined in NADs by using networkManager.
- for `syncNAD`
  As multiple NADs may define networks with the same name, these networks should also have the same config to be considered compatible. If an incompatible network change happens on NAD update, we can
  1. Re-create network with the same name but different config, if it is not used by any other NAD
  2. Return an error AND clean up NAD from the old network otherwise